File: backend/Vaveyla.Api/Services/ProfanityValidator.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Vaveyla.Api.Services;

public static class ProfanityValidator
{
    private static readonly string[] Blacklist =
    [
        "amk", "aq", "orospu", "piç", "pic", "sik", "sikeyim", "siktir", "göt", "got",
        "yarrak", "yarrrak", "mal", "salak", "aptal", "gerizekali", "gerizekalı",
        "kahpe", "pezevenk", "ibne", "oc", "oç", "anan", "ananı", "ananizi",
        "serefsiz", "şerefsiz", "haysiyetsiz", "it", "köpek", "kopek",
    ];

    public static bool ContainsProfanity(string? text)
    {
        if (string.IsNullOrWhiteSpace(text))
        {
            return false;
        }

        var normalized = Normalize(text);
        foreach (var word in Blacklist)
        {
            if (normalized.Contains(Normalize(word), StringComparison.Ordinal))
            {
                return true;
            }
        }

        return false;
    }

    public static void EnsureClean(string? text, string fieldLabel = "Metin")
    {
        if (ContainsProfanity(text))
        {
            throw new InvalidOperationException(
                $"{fieldLabel} uygunsuz ifadeler içeremez. Lütfen saygılı bir dil kullanın.");
        }
    }

    private static string Normalize(string input)
    {
        var lower = input.ToLower(new CultureInfo("tr-TR"));
        lower = lower
            .Replace('ı', 'i')
            .Replace('ğ', 'g')
            .Replace('ü', 'u')
            .Replace('ş', 's')
            .Replace('ö', 'o')
            .Replace('ç', 'c');
        return Regex.Replace(lower, @"[^a-z0-9]+", " ");
    }
}

File: lib/core/utils/profanity_validator.dart
/// Türkçe küfür/hakaret/argo için temel metin kontrolü (büyük-küçük harf duyarsız).
class ProfanityValidator {
  ProfanityValidator._();

  static const List<String> _blacklist = [
    'amk',
    'aq',
    'orospu',
    'piç',
    'pic',
    'sik',
    'sikeyim',
    'siktir',
    'göt',
    'got',
    'yarrak',
    'mal',
    'salak',
    'aptal',
    'gerizekali',
    'gerizekalı',
    'kahpe',
    'pezevenk',
    'ibne',
    'oc',
    'oç',
    'anan',
    'ananı',
    'ananizi',
    'serefsiz',
    'şerefsiz',
    'haysiyetsiz',
    'köpek',
    'kopek',
  ];

  static const String profanityWarning =
      'Mesajınız uygunsuz ifadeler içeriyor. Lütfen saygılı bir dil kullanın.';

  static bool containsProfanity(String? text) {
    if (text == null || text.trim().isEmpty) return false;
    final normalized = _normalize(text);
    for (final word in _blacklist) {
      if (normalized.contains(_normalize(word))) {
        return true;
      }
    }
    return false;
  }

  static String? validateOrNull(String? text) {
    return containsProfanity(text) ? profanityWarning : null;
  }

  static String _normalize(String input) {
    var s = input.toLowerCase();
    const map = {
      'ı': 'i',
      'ğ': 'g',
      'ü': 'u',
      'ş': 's',
      'ö': 'o',
      'ç': 'c',
      'İ': 'i',
      'Ğ': 'g',
      'Ü': 'u',
      'Ş': 's',
      'Ö': 'o',
      'Ç': 'c',
    };
    map.forEach((from, to) {
      s = s.replaceAll(from, to);
    });
    return s.replaceAll(RegExp(r'[^a-z0-9]+'), ' ');
  }
}
